lib/model/message_list.dart

```diff
@@ -485,6 +505,84 @@ class MessageListView with ChangeNotifier, _MessageSequence {
     }
   }
 
+  void _messagesMovedInternally(List<int> messageIds) {
+    for (final messageId in messageIds) {
+      if (_findMessageWithId(messageId) != -1) {
+        _reprocessAll();
+        notifyListeners();
+        return;
+      }
+    }
+  }
+
+  void _messagesMovedIntoNarrow() {
+    // If there are some messages we don't have in [MessageStore], and they
+    // occur later than the messages we have here, then we just have to
+    // re-fetch from scratch.  That's always valid, so just do that always.
+    // TODO in cases where we do have data to do better, do better.
+    _reset();
+    notifyListeners();
+    fetchInitial();
+  }
+
+  void _messagesMovedFromNarrow(
+    List<int> messageIds,
+    {TopicNarrow? newNarrow, PropagateMode? propagateMode}
+  ) {
+    if (_removeMessagesById(messageIds)) {
+      if (propagateMode == PropagateMode.changeAll
+          || propagateMode == PropagateMode.changeLater) {
+        narrow = newNarrow!;
```

Comment:
Separately: I think the overall behavior is going to be pretty broken as of this commit — the `MessageListView` changes what narrow it's tracking, but the associated `MessageListPage` is still about the old narrow. So the app bar still shows the old narrow, and the compose box will still send to the old narrow, even though the messages you're reading (including even any new messages that come in) are in the new narrow.

I think that can be fixed by just squashing the two commits together. They're both pretty small changes outside their tests.